subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpClientConfigurer.java

```diff
@@ -159,40 +159,54 @@ private void configureProxy(HttpClientBuilder builder, CredentialsProvider crede
+                    authentication.addHost(proxy.host, proxy.port);
+                    useCredentials(credentialsProvider, Collections.singleton(authentication));
                 }
             }
         }
         builder.setRoutePlanner(new SystemDefaultRoutePlanner(ProxySelector.getDefault()));
     }
 
-    private void useCredentials(CredentialsProvider credentialsProvider, String host, int port, Collection<? extends Authentication> authentications) {
+    private void useCredentials(CredentialsProvider credentialsProvider, Collection<? extends Authentication> authentications) {
         for (Authentication authentication : authentications) {
+            AuthenticationInternal authenticationInternal = (AuthenticationInternal) authentication;
+
             String scheme = getAuthScheme(authentication);
-            org.gradle.api.credentials.Credentials credentials = ((AuthenticationInternal) authentication).getCredentials();
+            org.gradle.api.credentials.Credentials credentials = authenticationInternal.getCredentials();
 
-            if (credentials instanceof HttpHeaderCredentials) {
-                HttpHeaderCredentials httpHeaderCredentials = (HttpHeaderCredentials) credentials;
-                Credentials httpCredentials = new HttpClientHttpHeaderCredentials(httpHeaderCredentials.getName(), httpHeaderCredentials.getValue());
-                credentialsProvider.setCredentials(new AuthScope(host, port, AuthScope.ANY_REALM, scheme), httpCredentials);
+            Collection<AuthenticationInternal.HostAndPort> hostsForAuthentication = authenticationInternal.getHostsForAuthentication();
+            assert !hostsForAuthentication.isEmpty() : "Credentials and authentication required for a HTTP repository, but no hosts were defined for the authentication?";
 
-                LOGGER.debug("Using {} for authenticating against '{}:{}' using {}", httpHeaderCredentials, host, port, scheme);
-            } else if (credentials instanceof PasswordCredentials) {
-                PasswordCredentials passwordCredentials = (PasswordCredentials) credentials;
+            for (AuthenticationInternal.HostAndPort hostAndPort : hostsForAuthentication) {
+                String host = hostAndPort.getHost();
+                int port = hostAndPort.getPort();
 
-                if (authentication instanceof AllSchemesAuthentication) {
-                    NTLMCredentials ntlmCredentials = new NTLMCredentials(passwordCredentials);
-                    Credentials httpCredentials = new NTCredentials(ntlmCredentials.getUsername(), ntlmCredentials.getPassword(), ntlmCredentials.getWorkstation(), ntlmCredentials.getDomain());
-                    credentialsProvider.setCredentials(new AuthScope(host, port, AuthScope.ANY_REALM, AuthSchemes.NTLM), httpCredentials);
+                assert host != null : "HTTP credentials and authentication require a host scope to be defined as well";
```

Comment:
Do you want to add a check that `port != -1` here?

Reply:
The port is going to be `-1` a lot of the time because the URL doesn't specify the port explicitly.  We would need to check the protocol and assume the right default (80 for http and 443 for https).  That seems doable, but I think it would raise the danger of breaking someone that has their repository setup to redirect http to https and only use the http URL with authentication?

Are hosts usually distinguishable by port number for these kinds of auth schemes?

Reply:
I think your assessment is appropriate here.

> Are hosts usually distinguishable by port number for these kinds of auth schemes?

I don't understand the question.

Reply:
I mean, I think you *could* have different sets of credentials for `example.com:1234` and `example.com:5678`, but if we didn't allow that for now, would that be a big deal?